nexus/src/db/subnet_allocation.rs

```diff
@@ -490,142 +305,588 @@ impl QueryFragment<Pg> for FilterConflictingVpcSubnetRangesQueryValues {
+fn push_ensure_unique_vpc_expression(
+    mut out: AstPass<Pg>,
+    vpc_id: &Uuid,
+    instance_id: &Uuid,
+) -> diesel::QueryResult<()> {
+    use db::schema::network_interface::dsl;
+
+    out.push_sql("CAST(IF(COALESCE((SELECT ");
+    out.push_identifier(dsl::vpc_id::NAME)?;
+    out.push_sql(" FROM ");
+    dsl::network_interface.from_clause().walk_ast(out.reborrow())?;
+    out.push_sql(" WHERE ");
+    out.push_identifier(dsl::time_deleted::NAME)?;
+    out.push_sql(" IS NULL AND ");
+    out.push_identifier(dsl::instance_id::NAME)?;
+    out.push_sql(" IS NOT NULL AND ");
+    out.push_identifier(dsl::instance_id::NAME)?;
+    out.push_sql(" = ");
+    out.push_bind_param::<sql_types::Uuid, Uuid>(instance_id)?;
+    out.push_sql(" LIMIT 1), ");
+    out.push_bind_param::<sql_types::Uuid, Uuid>(vpc_id)?;
+    out.push_sql(") = ");
+    out.push_bind_param::<sql_types::Uuid, Uuid>(vpc_id)?;
+    out.push_sql(", ");
+
+    // NOTE: This bind-parameter is intentionally a string, rather than a UUID.
+    //
+    // This query relies on the fact that it generates a parsing error in the
+    // case where there is an interface attached to a VPC that's _different_
+    // from the VPC of the candidate interface. This is so that we can
+    // distinguish this error case from the one where there is no IP address
+    // available.
+    //
+    // To do that, we generate a query like:
+    //
+    // ```
+    // CAST(IF(<instance VPC is the same>, '<vpc_id>', '') AS UUID)
+    // ```
+    //
+    // That empty-string cannot be cast to a UUID, so we get a parsing error,
+    // but only if the condition _succeeds_. It's not evaluated otherwise.
+    //
+    // However, if we push this parameter as a UUID explicitly, the database
+    // looks at the parts of the `IF` statement, and tries to make them a common
+    // type, a UUID. That's the exact error we're trying to produce, but it's
+    // evaluated too early. So we ensure both are strings here, and then ask the
+    // DB to cast them after that condition is evaluated.
+    out.push_bind_param::<sql_types::Text, String>(&vpc_id.to_string())?;
+    out.push_sql(", '') AS UUID)");
+    Ok(())
+}
+
+/// Push a subquery that selects the next available IP address from a subnet.
+///
+/// This adds a subquery like:
+///
+/// ```sql
+/// SELECT
+///     <network_address> + address_offset AS ip
+/// FROM
+///     generate_series(5, <last_address_offset>) AS address_offset
+/// LEFT OUTER JOIN
+///     network_interface
+/// ON
+///     (subnet_id, ip, time_deleted IS NULL) =
+///     (<subnet_id, <network_address> + address_offset, TRUE)
+/// WHERE ip IS NULL LIMIT 1
+/// ```
+///
+/// This is a linear, sequential scan for an IP from the subnet that's not yet
+/// been allocated. We'd ultimately like a better-performing allocation
+/// strategy; for example, we might be able to keep the lowest unallocated
+/// address for each subnet, and atomically return and increment that.
```

Comment:
This seems like a shortcut, but IMO a reasonable one considering that we're trying to move relatively quickly.

Reply:
Yep totally. I think Dave and I discussed something along these lines, and I wanted to write that down in here so we have a place to start if we decide its necessary.